File: app/services/compliance.py
from sqlalchemy.orm import Session

from app.models.contracts import Contract
from app.models.obligations import Obligation


def calculate_contract_compliance(
    db: Session,
    contract_id: int
):
    contract = db.query(Contract).filter(
        Contract.id == contract_id
    ).first()

    if not contract:
        return None

    obligations = db.query(Obligation).filter(
        Obligation.contract_id == contract_id
    ).all()

    total = len(obligations)

    if total == 0:
        return {
            "contract_id": contract_id,
            "compliance_status": "Pending",
            "compliance_score": 0,
            "total_obligations": 0,
            "completed_obligations": 0,
            "pending_obligations": 0,
            "delayed_obligations": 0,
            "overdue_obligations": 0,
            "risk_level": "Low"
        }

    completed = sum(
        1 for o in obligations
        if o.status == "Completed"
    )

    pending = sum(
        1 for o in obligations
        if o.status == "Pending"
    )

    delayed = sum(
        1 for o in obligations
        if o.status == "Delayed"
    )

    overdue = sum(
        1 for o in obligations
        if o.status == "Overdue"
    )

    compliance_score = round(
        (completed / total) * 100,
        2
    )

    # Compliance status
    if overdue > 0:
        compliance_status = "Non-Compliant"
    elif delayed > 0:
        compliance_status = "Delayed"
    elif pending > 0:
        compliance_status = "Pending"
    elif completed == total:
        compliance_status = "Compliant"
    else:
        compliance_status = "Pending"

    # Risk level
    if overdue >= 2:
        risk_level = "High"
    elif overdue == 1:
        risk_level = "Medium"
    else:
        risk_level = "Low"

    return {
        "contract_id": contract_id,
        "compliance_status": compliance_status,
        "compliance_score": compliance_score,
        "total_obligations": total,
        "completed_obligations": completed,
        "pending_obligations": pending,
        "delayed_obligations": delayed,
        "overdue_obligations": overdue,
        "risk_level": risk_level
    }


def get_all_compliance(db: Session):
    contracts = db.query(Contract).all()

    results = []

    for contract in contracts:
        compliance = calculate_contract_compliance(
            db,
            contract.id
        )

        results.append({
            "contract_id": contract.id,
            "contract_number": contract.contract_number,
            "compliance_status": compliance["compliance_status"],
            "compliance_score": compliance["compliance_score"],
            "overdue_obligations": compliance["overdue_obligations"],
            "risk_level": compliance["risk_level"]
        })

    return results


def get_compliance_summary(db: Session):
    records = get_all_compliance(db)

    return {
        "total_contracts": len(records),
        "compliant_contracts": sum(
            1 for r in records
            if r["compliance_status"] == "Compliant"
        ),
        "pending_contracts": sum(
            1 for r in records
            if r["compliance_status"] == "Pending"
        ),
        "delayed_contracts": sum(
            1 for r in records
            if r["compliance_status"] == "Delayed"
        ),
        "non_compliant_contracts": sum(
            1 for r in records
            if r["compliance_status"] == "Non-Compliant"
        ),
        "high_risk_contracts": sum(
            1 for r in records
            if r["risk_level"] == "High"
        )
    }


def get_non_compliant_contracts(db: Session):
    records = get_all_compliance(db)

    return [
        r for r in records
        if r["compliance_status"] == "Non-Compliant"
    ]


def get_high_risk_contracts(db: Session):
    records = get_all_compliance(db)

    return [
        r for r in records
        if r["risk_level"] == "High"
    ]

File: contractiq_frontend/src/app/compliance/compliance.ts
import { Component, OnInit, ChangeDetectorRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MatCardModule } from '@angular/material/card';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatButtonModule } from '@angular/material/button';
import { MatTableModule } from '@angular/material/table';
import { MatSelectModule } from '@angular/material/select';

import {
  ComplianceService,
  ComplianceRecord,
  ComplianceSummary
} from '../services/compliance';

@Component({
  selector: 'app-compliance',
  standalone: true,
  imports: [
    CommonModule,
    FormsModule,
    MatCardModule,
    MatFormFieldModule,
    MatInputModule,
    MatButtonModule,
    MatTableModule,
    MatSelectModule
  ],
  templateUrl: './compliance.html',
  styleUrl: './compliance.css'
})
export class Compliance implements OnInit {

  records: ComplianceRecord[] = [];
  filteredRecords: ComplianceRecord[] = [];

  summary: ComplianceSummary | null = null;

  loading = false;

  errorMessage = '';
  successMessage = '';

  searchText = '';
  selectedStatus = '';
  selectedRisk = '';

  selectedRecord: ComplianceRecord | null = null;
  history: any[] = [];
  historyLoading = false;

  displayedColumns = [
    'contract_number',
    'compliance_status',
    'compliance_score',
    'total_obligations',
    'completed_obligations',
    'pending_obligations',
    'overdue_obligations',
    'risk_level',
    'actions'
  ];

  statuses = [
    'Compliant',
    'Pending',
    'Delayed',
    'Non-Compliant',
    'High Risk'
  ];

  riskLevels = [
    'Low',
    'Medium',
    'High'
  ];

  constructor(
    private complianceService: ComplianceService,
    private cdr: ChangeDetectorRef
  ) {}

  ngOnInit(): void {
    this.loadCompliance();
  }

  loadCompliance(): void {
    this.loading = true;
    this.errorMessage = '';

    this.complianceService.getComplianceRecords().subscribe({
      next: (data) => {
        this.records = data;
        this.applyFilters();
        this.loading = false;
        this.cdr.markForCheck();
      },
      error: (error) => {
        console.error('Compliance API error:', error);
        this.errorMessage = 'Unable to load compliance records.';
        this.loading = false;
        this.cdr.markForCheck();
      }
    });

    this.complianceService.getComplianceSummary().subscribe({
      next: (data) => {
        this.summary = data;
        this.cdr.markForCheck();
      },
      error: (error) => {
        console.error('Compliance summary API error:', error);
      }
    });
  }

  applyFilters(): void {
    const search = this.searchText.toLowerCase().trim();

    this.filteredRecords = this.records.filter(record => {

      const matchesSearch =
        !search ||
        record.contract_number.toLowerCase().includes(search) ||
        String(record.contract_id).includes(search);

      const matchesStatus =
        !this.selectedStatus ||
        record.compliance_status === this.selectedStatus;

      const matchesRisk =
        !this.selectedRisk ||
        record.risk_level === this.selectedRisk;

      return matchesSearch && matchesStatus && matchesRisk;
    });
  }

  clearFilters(): void {
    this.searchText = '';
    this.selectedStatus = '';
    this.selectedRisk = '';
    this.applyFilters();
  }

  viewDetails(record: ComplianceRecord): void {
    this.selectedRecord = record;
    this.history = [];
    this.historyLoading = true;
    this.errorMessage = '';

    this.complianceService
      .getComplianceHistory(record.contract_id)
      .subscribe({
        next: (data) => {
          this.history = data;
          this.historyLoading = false;
          this.cdr.markForCheck();
        },
        error: (error) => {
          console.error('Compliance history error:', error);
          this.historyLoading = false;
          this.errorMessage =
            'Unable to load compliance history.';
          this.cdr.markForCheck();
        }
      });
  }

  closeDetails(): void {
    this.selectedRecord = null;
    this.history = [];
  }

  getScoreClass(score: number): string {
    if (score >= 80) {
      return 'score-good';
    }

    if (score >= 50) {
      return 'score-medium';
    }

    return 'score-low';
  }

  getStatusClass(status: string): string {
    return status
      .toLowerCase()
      .replace(/\s+/g, '-');
  }

  getRiskClass(risk: string): string {
    return risk.toLowerCase();
  }
}
